scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/DTLSSession.java

```diff
@@ -564,6 +573,27 @@ synchronized public String getWriteStateCipher() {
 		return writeState.getCipherSuite().name();
 	}
 
+	/**
+	 * Set parameter available. Enables {@link #getParameter()} to return the
+	 * handshake parameter.
+	 */
+	public void setParameterAvailable() {
+		parameterAvailable = true;
```

Comment:
After trying out some variants I found that it is the clearest way, to use this pattern. 
The `CipherSuite cipherSuite = CipherSuite.TLS_NULL_WITH_NULL_NULL` is initialized, and so, I didn't want to use it.

Reply:
I feel that it would be better to check if we really have the information on the session instead of using a flag which say that we should have set this information. Maybe less error prone ?

I would use : 
```
CipherSuite cipherSuite = null;
Boolean receiveRawPublicKey = null;
```
to express parameters are not set.

Reply:
I had this also, but it makes it more complex in my opinion.
With ciphersuite = null; the unit test fails.
And to setup Boolean receiveRawPublicKey you must take care, that is setup also, if no extensions are used. So my conclusion was, to use a explicit flag.

Reply:
So for receiveRawPublicKey we can use default value false ?
And so once we have the cipher suite we can consider parameter is set, right ?

Reply:
> So for receiveRawPublicKey we can use default value false ?
And so once we have the cipher suite we can consider parameter is set, right ?

I don't know, if your claims are right. And again, FMPOV this variant, with an explicit  parameterAvailable is the clearest way for me.

Reply:
> And again, FMPOV this variant, with an explicit parameterAvailable is the clearest way for me.

I do not understand why adding a parameters which says "that all needed parameters are set" is clearer than just set the expected parameter and have a function which return if this expected parameter are set or not ? It's feel to me error prone as you can face inconsistent state.

The test issue sounds not really a good argument to me, so I try to figure out your point about receiveRawPublicKey.

Reply:
You made a proposal, https://github.com/eclipse/californium/pull/724#discussion_r215187773, but that didn't work. So, at least on the first impression, that solution can't be clear, because it failed from pitfalls caused by the current implementation.

> set the expected parameter and have a function which return if this expected parameter are set or not ? 

Because some stuff is set from the extensions (which are optional) and so you must also take care of the case, when the extension is not used/sent. That makes it more complex and error prone.

In the future we will see, which uncovered race condition we will be faced, Then we will see, which variant would have been better :-).

Reply:
> Because some stuff is set from the extensions (which are optional) and so you must also take care of the case, when the extension is not used/sent. That makes it more complex and error prone.

I don't know which stuff you are talking about, it seems to me that the only information which are concerned by this PR is cipherSuite and receiveRawPublicKey ?

Anyway, I'm still feel this pattern is not so good, but I didn't spend so many time on this. But generally the next step is going in an endless discussion then I will provide a PR, then you don't like this and we will get stuck and finally we will integrate this PR. :D

So let's go directly to the last step ;)

Reply:
So, "last step".